### ISSUE TITLE
How caffe calculate total loss when training faster rcnn by Approximate joint training way?

### ISSUE THREAD

**Issue by XiaotaoChen** [BVLC/caffe#5820]
Hi,
   recently, i read faster rcnn's source code. i have two questions as follows,

1. **how caffe calculate total loss among rpn_cls_loss, rpn_loss_bbox,loss_cls, loss_bbox?**

   i haven't found the 'sum loss' code in faster rcnn's source code. Can caffe auto find the multiple loss in the defined net  and sum them weight by each loss_weight? How is it work? Can anyone explain it and tell me this part code in caffe cource?

2. **how the network do the backwards according to the total_loss?**

   after got the total loss. there are four branches(layers which top-out is rpn_cls_loss, rpn_loss_bbox, loss_cls, loss_bbox) needed to do backwards ?  Firstly, how to allocate the total loss to four branches ? is each branch' loss= total-loss/4 and then to do backwards independently ? Secondly, the four branches are above on VGG16. So if The four branches do backwards independently, The paramaters in VGG16 will be updated four times . Is my guess correct ?

**Comment by shohan6** [comment]
By convention, Caffe layer types with the suffix Loss contribute to the loss function, but other layers are assumed to be purely used for intermediate computations. However, any layer can be used as a loss by adding a field loss_weight: <float> to a layer definition for each top blob produced by the layer. Layers with the suffix Loss have an implicit loss_weight: 1 for the first top blob (and loss_weight: 0 for any additional tops); other layers have an implicit loss_weight: 0 for all tops. So, the above SoftmaxWithLoss layer could be equivalently written as:

`layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "pred"
  bottom: "label"
  top: "loss"
  loss_weight: 1
}`
However, any layer able to backpropagate may be given a non-zero loss_weight, allowing one to, for example, regularize the activations produced by some intermediate layer(s) of the network if desired. For non-singleton outputs with an associated non-zero loss, the loss is computed simply by summing over all entries of the blob.

The final loss in Caffe, then, is computed by summing the total weighted loss over the network, as in the following pseudo-code:

`loss := 0
for layer in layers:
  for top, loss_weight in layer.tops, layer.loss_weights:
    loss += loss_weight * sum(top)`

see [http://caffe.berkeleyvision.org/tutorial/loss.html](here) for details

**Comment by Noiredd** [comment]
Please post usage, installation, or modeling questions, or other requests for help to the [caffe-users list][1] instead of Issues. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe. Please read the [guidelines for contributing][2] before submitting an issue or a pull request.

 [1]: https://groups.google.com/forum/#!forum/caffe-users
 [2]: https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md
